Update BF obsoletion timeline (#4)
GrabYourPitchforks/designs: accepted/2020/better-obsoletion/binaryformatter-obsoletion.md

### .NET 7 (Nov 2022)

> - All first-party dotnet org code bases continue migration away from `BinaryFormtter`

Reviewer: ```suggestion
- All first-party dotnet org code bases continue migration away from `BinaryFormatter`
```

> - APIs previously marked obsolete (_as warning_) in .NET 5 are marked obsolete (_as error_)

Reviewer: ```suggestion
- References to `BinaryFormatter` APIs marked obsolete as _warnings_ in .NET 5 now result in build _errors_ by default in .NET 7+
- Build _errors_ can be downgraded back to build _warnings_ by setting `<EnableUnsafeBinaryFormatterSerialization>true</EnableUnsafeBinaryFormatterSerialization>`
```

### .NET 8 (Nov 2023)

- All first-party dotnet org code bases complete migration away from `BinaryFormatter`
- `BinaryFormatter` disabled by default across all project types
- Entirety of `BinaryFormatter` type obsolete _as error_
- Entirety of legacy serialization infrastructure marked obsolete _as warning_

> - No new `[Serializable]` types introduced (all target frameworks)

Reviewer: ```suggestion
- Entirety of `BinaryFormatter` type obsolete, with references producing _errors_ by default
- Entirety of legacy serialization infrastructure marked obsolete _as warning_
- No new `[Serializable]` types introduced (all target frameworks)
- Build _errors_ can be downgraded back to build _warnings_ by setting `<EnableUnsafeBinaryFormatterSerialization>true</EnableUnsafeBinaryFormatterSerialization>`
```

### .NET 9 (Nov 2024)

> - `BinaryFormatter` infrastructure removed from .NET

Reviewer: ```suggestion
- `BinaryFormatter` infrastructure removed from .NET
- `<EnableUnsafeBinaryFormatterSerialization>` feature switch removed, with no mechanism to re-enable `BinaryFormatter` references or usage
```

As part of this, dotnet org developers should audit the area of ownership for calls to `Type.GetType` and related APIs, as these could function as unrestricted deserialization equivalents. Any call sites which deal with untrusted data should clearly be annotated as such in code comments. If necessary, the owners should open issues to track removing the dangerous code patterns or documenting the API as dangerous. This will allow us to make more informed estimates of our libraries' own exposure to potential RCE attacks.

> ### APIs previously marked obsolete (_as warning_) in .NET 5 are marked obsolete (_as error_) (.NET 7)

Reviewer: ```suggestion
### APIs previously marked obsolete (_as warning_) in .NET 5 now produce build _errors_ by default (.NET 7)
```

> In .NET 5, we marked several `BinaryFormatter`-related APIs obsolete as warning (see [the breaking change notice](https://docs.microsoft.com/dotnet/core/compatibility/core-libraries/5.0/binaryformatter-serialization-obsolete)). In .NET 7, this set of APIs will be marked obsolete _as error_. This will cause compilation failures in applications which directly consume these APIs, and developers who call these APIs must take action in order to migrate their applications onto .NET 7.

Reviewer: ```suggestion
In .NET 5, we marked several `BinaryFormatter`-related APIs obsolete as warning (see [the breaking change notice](https://docs.microsoft.com/dotnet/core/compatibility/core-libraries/5.0/binaryformatter-serialization-obsolete)). In .NET 7, references to these APIs will produce build _errors_. This will cause compilation failures in applications which directly consume these APIs, and developers who call these APIs must take action in order to migrate their applications onto .NET 7.
```

> A back-compat switch will be made available for applications which cannot migrate away from `BinaryFormatter` in the .NET 7 timeframe. Details of this back-compat switch will be included in the breaking change notice which accompanies the release.

Reviewer: ```suggestion
An opt-in, back-compat switch is available for applications which cannot migrate away from `BinaryFormatter` in the .NET 7 timeframe. Setting `<EnableUnsafeBinaryFormatterSerialization>true</EnableUnsafeBinaryFormatterSerialization>` results in build _warnings_ instead of _errors_.
```

No first-party _dotnet_ org code base contains an enabled-by-default calls to `BinaryFormatter`. The call sites may remain in the library but must be guarded by an opt-in switch. This opt-in switch must be specified by _the app developer_, not by an "I'm serialized using `BinaryFormatter`" flag within the payload itself. All other payload formats must be written or read using a `BinaryFormatter` alternative.

Importantly, none of our project types (WinForms / WPF / ASP.NET / mobile / etc.) can require developers to opt back in to allowing `BinaryFormatter` for any greenfield development. Applications which require `BinaryFormatter` for app-compat reasons must require developers to perform an explicit opt-in gesture to re-enable these code paths.

### BinaryFormatter disabled by default across all project types (.NET 8)

Beginning with .NET 8.0, the "disable `BinaryFormatter`" feature switch is activated by default across all project types. Developers must perform an explicit opt-in gesture to re-enable these code paths. See the earlier feature switch section for more information.

### New obsoletions in .NET 8

> In .NET 8.0, the following APIs will be marked `[Obsolete]` _as error_. Applications which directly target these APIs will fail to compile. No back-compat switch will be provided.

Reviewer: ```suggestion
In .NET 8.0, the following APIs will be marked `[Obsolete]`, producing _errors_ by default. Applications which directly target these APIs will fail to compile. The opt-in, back-compat switch remains available.
```

This will also follow a "better obsoletion" mechanism so that applications can suppress the entire category of warnings if needed.

_(n.b. Obsoleting the `[Serializable]` and `[NonSerialized]` attributes may require additional thought. They are intricately tied to the legacy serialization infrastructure and cannot correctly be used outside of it, but there are many serializers in the wild which special-case only these two attributes and which don't understand the contract they have with the rest of the legacy serialization stack.)_

### No new \[Serializable\] types introduced (.NET 8)

In .NET 6, we had stopped attributing new types with `[Serializable]`, carving out an exception only for `Exception`-derived types in projects which targeted _netfx_ or _netstandard_.

Starting with .NET 8, we will no longer attribute _any_ new types with `[Serializable]`, voiding the previously granted exception. (Already-shipped types will retain their existing annotations.)

### BinaryFormatter infrastructure removed from .NET (.NET 9)

> In .NET 9.0, the entirety of the `BinaryFormatter` infrastructure will be removed from the product. Here, "`BinaryFormatter` infrastructure" refers to any API which was obsoleted (_as error_) as part of an earlier stage of this proposal, even if that API is not itself on the `BinaryFormatter` type. The implementation of all such APIs will be to throw `PlatformNotSupportedException`. The internal `BinaryFormatter` implementation will no longer exist in .NET in any form. Functionality may be moved to an unsupported out-of-band package.

Reviewer: ```suggestion
In .NET 9.0, the entirety of the `BinaryFormatter` infrastructure will be removed from the product. Here, "`BinaryFormatter` infrastructure" refers to any API which was obsoleted (_as error_) as part of an earlier stage of this proposal, even if that API is not itself on the `BinaryFormatter` type. The implementation of all such APIs will be to throw `NotSupportedException`. The internal `BinaryFormatter` implementation will no longer exist in .NET in any form, and the back-compat `EnableUnsafeBinaryFormatterSerialization` switch will be removed. Functionality may be moved to an unsupported out-of-band package.
```
